Edit Other Duties

Jdellis818 opened the pull request:
This PR allows the user to edit other duties on the teaching loads tab. To test, add a non-teaching activity. Then click on that activity in the table and change one of the fields. It should update successfully. Deleting a non-teaching load should work as well.

Jdellis818 committed:
Get modal to appear and populate

Jdellis818 committed:
Finish edit functionality for non teaching loads

bryantgeorge commented:
I think this branch contains all of the changes from #120 ? Should I review this, and approve both?

bryantgeorge commented:
It seems that you can only click the non-teaching loads to edit if the teacher also has a regular course (otherwise clicking the non-teaching load does nothing)

bryantgeorge commented:
This branch seems to cause blank values to be filled in with 0s. To replicate, make a course with a minimal number of fields filled in (prefix, number, section letter, room, days, duration, and instructor), then click on the section to edit, and it will have 0s in many of the previously blank fields, like so:

![image](https://user-images.githubusercontent.com/51130302/110029742-d6acfd80-7d02-11eb-95be-34a7a60f9787.png)

bryantgeorge commented:
I don't understand what makes some items have a number beside them while others don't:

![image](https://user-images.githubusercontent.com/51130302/110030021-32778680-7d03-11eb-8a2c-e4bce89210d4.png)

Is this intentional?

bryantgeorge commented:
Might be nice to switch the mouse cursor to the pointing hand icon when hovering over a clickable non-teaching load or section.

bryantgeorge commented:
If multiple semesters are selected, should that multiply the value of faculty hours? (This might be another judgement call)

bryantgeorge commented on client-course-schedulizer/src/utilities/services/validation/popover.ts:
```diff
@@ -33,41 +33,36 @@ NOTE: fields with default values are not check: semester and time. */
   facultyHours: number()
     .typeError("faculty hours must be a number")
-    .positive()
+    .integer()
```
I think we don't want the integer check here. Also, I think we want to leave the positive checks? I've been able to update courses to have a negative global max for example which seems illogical.

Jdellis818 replied:
Positive doesn't work because it doesn't allow 0 as a possible value. I added integer because I thought it was better than nothing, but now realize that that doesn't work either since there can be fractional loads. I'll remove it, but I think we should figure out how to add a custom function to allow only positive numbers or 0.

rpruim replied:
Can you write a .nonnegative()?

Jdellis818 replied:
After some research, it looks like .min(0) should do the trick

charkour replied:
:dab:

![](https://user-images.githubusercontent.com/33156025/110053347-3b7a4f00-7d27-11eb-8a63-acc9f5f85a4e.png)

bryantgeorge commented:
Sometimes I'll edit a section, and it will change the days that the section meets on (even though I didn't add any days). This seems to happen when I have multiple sections for one course.

bryantgeorge reviewed (changes requested):
Okay, the above comments cover the extent of the things I noticed in testing. This is a really cool feature overall!

Notes:
- I spent most of my time testing and less time reading the code since there are so many lines changed.
- If any of the bugs I noticed are present here but not in #120 then let me know.

Jdellis818 commented:
> This branch seems to cause blank values to be filled in with 0s. To replicate, make a course with a minimal number of fields filled in (prefix, number, section letter, room, days, duration, and instructor), then click on the section to edit, and it will have 0s in many of the previously blank fields, like so:
> 
> ![image](https://user-images.githubusercontent.com/51130302/110029742-d6acfd80-7d02-11eb-95be-34a7a60f9787.png)

Yeah, I noticed this issue and figured out what caused it. Since nearly all of the fields in the Course, Section, and Meeting interfaces are required, the empty objects that are originally assigned to each section when the CSV is read, contain values of 0 for all the `number` fields. We fixed this before this PR by clearing any 0 values and interpreting them as empty, but that's not a good long-term solution since we should allow 0 as a possible value. In the future, we'll need to make most of the fields in the schedule interfaces optional, but that should be addressed in a different PR since it has a lot of ramifications especially in writing to the csv. I'll make an issue for it.

Jdellis818 committed:
Remove hours from other duties cell

Jdellis818 committed:
Remove integer validation for load hours

Jdellis818 commented:
> If multiple semesters are selected, should that multiply the value of faculty hours? (This might be another judgement call)

I don't think so based on how the template non-teaching loads spreadsheet was structured

Jdellis818 committed:
Add min(0) to validation on numbers

Jdellis818 commented:
> It seems that you can only click the non-teaching loads to edit if the teacher also has a regular course (otherwise clicking the non-teaching load does nothing)

@bryantgeorge I can't seem to reproduce this. How did you get this bug to occur?

Jdellis818 committed:
Add pointer cursor for table body

Jdellis818 commented:
> Might be nice to switch the mouse cursor to the pointing hand icon when hovering over a clickable non-teaching load or section.

I added a pointer for the cursor when hovering over the table body. It's better than it was before, but stays as a pointer even for the cells that are not clickable. I think it will be somewhat difficult to detect whether the table cell is actually clickable or not. Do you think it's worth it to try and distinguish and switch the cursor between the pointer and normal?

Jdellis818 committed:
Fix changing days bug

bryantgeorge commented:
> > It seems that you can only click the non-teaching loads to edit if the teacher also has a regular course (otherwise clicking the non-teaching load does nothing)
> 
> @bryantgeorge I can't seem to reproduce this. How did you get this bug to occur?

I went to the teaching loads section, and added a teaching load for a professor which doesn't exist yet, so it looked something like:

![image](https://user-images.githubusercontent.com/51130302/110059173-b5173a80-7d31-11eb-95be-ea8d7dcd4ac0.png)

and I wasn't able to click `Research (2)` unless I added a section for Jane Doe to teach.

Jdellis818 committed:
Fix terms checkbox

bryantgeorge commented:
> > Might be nice to switch the mouse cursor to the pointing hand icon when hovering over a clickable non-teaching load or section.
> 
> I added a pointer for the cursor when hovering over the table body. It's better than it was before, but stays as a pointer even for the cells that are not clickable. I think it will be somewhat difficult to detect whether the table cell is actually clickable or not. Do you think it's worth it to try and distinguish and switch the cursor between the pointer and normal?

Something like this should work:
```diff
diff --git a/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx b/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx
index 4abac7e..599329c 100644
--- a/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx
+++ b/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx
@@ -213,8 +213,14 @@ export const FacultyLoads = () => {
                     // Loop over the rows cells
                     row.cells.map((cell) => {
                       // Apply the cell props
+                      const cellClass = cell
+                        .getCellProps()
+                        .key.toString()
+                        .match(/(CourseSections)|(otherDuties)/g)
+                        ? "change-cursor"
+                        : "";
                       return (
-                        <TableCell {...cell.getCellProps()}>
+                        <TableCell className={cellClass} {...cell.getCellProps()}>
                           {
                             // Render the cell contents
                             cell.render("Cell")
```

Then only change the cursor for the `change-cursor` class

Jdellis818 committed:
Make button more visible

Jdellis818 commented:
> > > Might be nice to switch the mouse cursor to the pointing hand icon when hovering over a clickable non-teaching load or section.
> > 
> > 
> > I added a pointer for the cursor when hovering over the table body. It's better than it was before, but stays as a pointer even for the cells that are not clickable. I think it will be somewhat difficult to detect whether the table cell is actually clickable or not. Do you think it's worth it to try and distinguish and switch the cursor between the pointer and normal?
> 
> Something like this should work:
> 
> ```diff
> diff --git a/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx b/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx
> index 4abac7e..599329c 100644
> --- a/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx
> +++ b/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx
> @@ -213,8 +213,14 @@ export const FacultyLoads = () => {
>                      // Loop over the rows cells
>                      row.cells.map((cell) => {
>                        // Apply the cell props
> +                      const cellClass = cell
> +                        .getCellProps()
> +                        .key.toString()
> +                        .match(/(CourseSections)|(otherDuties)/g)
> +                        ? "change-cursor"
> +                        : "";
>                        return (
> -                        <TableCell {...cell.getCellProps()}>
> +                        <TableCell className={cellClass} {...cell.getCellProps()}>
>                            {
>                              // Render the cell contents
>                              cell.render("Cell")
> ```
> 
> Then only change the cursor for the `change-cursor` class

Awesome, thanks!

Jdellis818 committed:
Update cursor with Bryant's suggestion

Jdellis818 commented:
@bryantgeorge Thanks so much for the great review! I'm sorry for marking this PR ready for review so early. There were a lot of bugs, but you did a great job testing and catching a bunch that I missed! I think I've addressed all the issues which you brought up. I still couldn't reproduce the bug for professors with no classes though, so let me know if you're still experiencing that issue.

charkour commented on client-course-schedulizer/src/utilities/services/facultyLoadsService.ts:
```diff
@@ -0,0 +1,279 @@
+  if (!sections.length) {
+    return null;
+  }
+  const [section] = sections;
```
firstSection

Jdellis818 committed:
Add checkbox change comment

Jdellis818 committed:
Add default props to checkbox group

charkour commented on client-course-schedulizer/src/utilities/services/facultyLoadsService.ts:
```diff
@@ -220,3 +214,66 @@ export const getCourseSectionMeetingFromCell = (
+  if (!sections.length) {
+    return null;
+  }
+  const [section] = sections;
```
Possibly use more explicit variables names here that the course is the non-teaching loads and section is an individual non-teaching load.

charkour commented on client-course-schedulizer/src/components/Tabs/FacultyLoads/UpdateNonTeachingLoadModalPagination/UpdateNonTeachingLoadModalPagination.tsx:
```diff
@@ -0,0 +1,93 @@
+        <AddNonTeachingLoadPopover values={modalValues} />
+      </ModalPagination>
+    );
+  },
```
This is similar to the UpdateSectionModalPagination. I'm trying to think if it would be worth it to make one component more generic, or if it's fine to leave these separate but similar.

charkour replied:
I'm trying to see if we can combine them.

charkour committed:
combine popover modal logic

bryantgeorge commented:
Not a big deal, but this will make it so that blank cells won't look clickable:

Applied in 06c2eca

```diff
diff --git a/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx b/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx
index d9cb67f..f965d2d 100644
--- a/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx
+++ b/client-course-schedulizer/src/components/Tabs/FacultyLoads/FacultyLoads.tsx
@@ -133,12 +133,14 @@ export const FacultyLoads = () => {
                       // Loop over the rows cells
                       row.cells.map((cell) => {
                         // Apply the cell props
-                        const cellClass = cell
-                          .getCellProps()
-                          .key.toString()
-                          .match(/(CourseSections)|(otherDuties)/g)
-                          ? "change-cursor"
-                          : "";
+                        const cellClass =
+                          cell.value &&
+                          cell
+                            .getCellProps()
+                            .key.toString()
+                            .match(/(CourseSections)|(otherDuties)/g)
+                            ? "change-cursor"
+                            : "";
                         return (
                           <TableCell
                             className={cellClass}
```

bryantgeorge commented:
> I still couldn't reproduce the bug for professors with no classes though, so let me know if you're still experiencing that issue.

I wasn't able to reproduce this when I tried today, so seems that it's been fixed (or I was just doing something wrong in the first place)

bryantgeorge committed:
make empty fields not look clickable

charkour committed:
update documentation

charkour committed:
Merge pull request #144 from senior-knights/feature/popover

combine popover modal logic

charkour reviewed (approved):
This is good to be merged after some better documentation is added.

Jdellis818 committed:
add better variable names

Jdellis818 committed:
Merge branch 'feature/edit-other-duties' of https://github.com/senior-knights/course-schedulizer into feature/edit-other-duties
